main/src/ca/uwaterloo/flix/language/phase/unification/Zhegalkin.scala

```diff
@@ -0,0 +1,171 @@
+    mkXor(ZhegalkinExpr(ZUniverse, Nil), a)
+
+  //
+  // (c1 ? t11 ? t12 ? ... ? t1n) n (c2 ? t21 ? t22 ? ... ? t2m)
+  //   =   (c1  n (c2 ? t21 ? t22 ? ... ? t2m)
+  //     ? (t11 n (c2 ? t21 ? t22 ? ... ? t2m)
+  //     ? (t12 n (c2 ? t21 ? t22 ? ... ? t2m)
+  //
+  private def mkInter(z1: ZhegalkinExpr, z2: ZhegalkinExpr): ZhegalkinExpr = z1 match {
+    case ZhegalkinExpr(c1, ts1) =>
+      val zero = mkInterConstantExpr(c1, z2)
+      ts1.foldLeft(zero) {
+        case (acc, z) => mkXor(acc, mkInterTermExpr(z, z2))
+      }
+  }
+
+  //
+  // c n (c2 n x1 n x2 n ... n xn) = (c n c2) n x1 n x2 n ... n xn)
+  //
+  private def mkInterConstantTerm(c: ZhegalkinConstant, t: ZhegalkinTerm): ZhegalkinTerm = t match {
+    case ZhegalkinTerm(c2, vars) =>
+      ZhegalkinTerm(c.inter(c2), vars)
+  }
+
+  //
+  // c n (c2 ? t21 ? t22 ? ... ? t2m) = (c n c2) ? t21 ? t22 ? ... ? t2m
+  //
+  private def mkInterConstantExpr(c: ZhegalkinConstant, z: ZhegalkinExpr): ZhegalkinExpr = z match {
+    case ZhegalkinExpr(c2, terms) =>
+      val ts = terms.map(t => mkInterConstantTerm(c, t))
+      mkZhegalkinExpr(c.inter(c2), ts)
+  }
+
+  //
+  // t n (c2 ? t1 ? t2 ? ... ? tn) = (t n c2) ? (t n t1) ? (t n t2) ? ... ? (t n tn)
+  //
+  private def mkInterTermExpr(t: ZhegalkinTerm, z: ZhegalkinExpr): ZhegalkinExpr = z match {
+    case ZhegalkinExpr(c2, terms) =>
+      val zero: ZhegalkinExpr = mkZhegalkinExpr(ZEmpty, List(mkInterConstantTerm(c2, t)))
+      terms.foldLeft(zero) {
+        case (acc, t2) => mkXor(acc, mkZhegalkinExpr(ZEmpty, List(mkInterTermTerm(t, t2))))
+      }
+  }
+
+  // (c1 n x11 n x12 n ... n x1n) n (c2 n x21 n x22 n ... n x2m)
+  private def mkInterTermTerm(t1: ZhegalkinTerm, t2: ZhegalkinTerm): ZhegalkinTerm = (t1, t2) match {
+    case (ZhegalkinTerm(c1, vars1), ZhegalkinTerm(c2, vars2)) =>
+      ZhegalkinTerm(c1.inter(c2), vars1 ++ vars2)
+  }
+
+  /** Returns the union of the two Zhegalkin expressions. */
+  private def mkUnion(a: ZhegalkinExpr, b: ZhegalkinExpr): ZhegalkinExpr = {
+    /** a ? b = a ? b ? (a n b) */
+    mkXor(mkXor(a, b), mkInter(a, b))
+  }
+
+  /**
+    * Returns the given set formula as a Zhegalkin polynomial.
+    */
+  def toZhegalkin(f: SetFormula): ZhegalkinExpr = f match {
+    case SetFormula.Univ => ZhegalkinExpr(ZUniverse, Nil)
+    case SetFormula.Empty => ZhegalkinExpr(ZEmpty, Nil)
+    case Cst(c) => ZhegalkinExpr(ZEmpty, List(ZhegalkinTerm(ZUniverse, SortedSet(ZhegalkinVar(c))))) // We treat uninterpreted constants as vars.
+    case Var(x) => ZhegalkinExpr(ZEmpty, List(ZhegalkinTerm(ZUniverse, SortedSet(ZhegalkinVar(x)))))
+    case ElemSet(s) =>
+      ZhegalkinExpr(ZhegalkinConstant(CofiniteIntSet.mkSet(s)), Nil)
+    case Compl(f) => zmkCompl(toZhegalkin(f))
+    case Inter(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other) =>
+      val terms = SetFormula.subformulasOf(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other).toList
+      val polys = terms.map(toZhegalkin)
```

Comment:
Definitely non-empty?

Reply:
What do you mean? And I don't know. I don't fully understand Jonathans representation.

Reply:
Oh, your concern is the call to reduce. Yeah, I just don't know Jonathans rep. But in any case, if this code is ever to be used SetFormula would be deprecated and then the above code would go away.

Reply:
For now this is just an-experiment-in-progress.